dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/changelog/ChangeLogServiceTest.java

```diff
@@ -0,0 +1,311 @@
+    assertNotNull(dataElement);
+
+    List<EventChangeLog> changeLogs = changeLogService.getEventChangeLog(UID.of("QRYjLTiJTrA"));
+
+    assertAll(
+        () -> {
+          validateNonValueElements(1, changeLogs, importUser, dataElementUid);
+          assertEquals("15", changeLogs.get(0).change().dataValue().currentValue());
+          assertNull(changeLogs.get(0).change().dataValue().previousValue());
+        });
+  }
+
+  @Test
+  void shouldReturnChangeLogsWhenDataValueIsDeleted() throws NotFoundException, IOException {
+    injectSecurityContextUser(manager.get(User.class, "M5zQapPyTZI"));
+
+    Event event = manager.get(Event.class, "QRYjLTiJTrA");
+    assertNotNull(event);
+    String dataElementUid = event.getEventDataValues().iterator().next().getDataElement();
+    DataElement dataElement = manager.get(DataElement.class, dataElementUid);
+    assertNotNull(dataElement);
+
+    TrackerObjects trackerObjects = fromJson("tracker/event_and_enrollment.json");
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+
+    List<EventChangeLog> changeLogs = changeLogService.getEventChangeLog(UID.of("QRYjLTiJTrA"));
+
+    assertAll(
+        () -> {
+          validateNonValueElements(2, changeLogs, importUser, dataElementUid);
+          assertNull(changeLogs.get(0).change().dataValue().currentValue());
+          assertEquals("15", changeLogs.get(0).change().dataValue().previousValue());
+          assertEquals("15", changeLogs.get(1).change().dataValue().currentValue());
+          assertNull(changeLogs.get(1).change().dataValue().previousValue());
+        });
+  }
+
+  @Test
+  void shouldNotUpdateChangeLogsWhenDataValueIsDeletedTwiceInARow()
+      throws NotFoundException, IOException {
+    injectSecurityContextUser(manager.get(User.class, "M5zQapPyTZI"));
+
+    Event event = manager.get(Event.class, "QRYjLTiJTrA");
+    assertNotNull(event);
+    String dataElementUid = event.getEventDataValues().iterator().next().getDataElement();
+    DataElement dataElement = manager.get(DataElement.class, dataElementUid);
+    assertNotNull(dataElement);
+
+    TrackerObjects trackerObjects = fromJson("tracker/event_and_enrollment.json");
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+
+    List<EventChangeLog> changeLogs = changeLogService.getEventChangeLog(UID.of("QRYjLTiJTrA"));
+
+    assertAll(
+        () -> {
+          validateNonValueElements(2, changeLogs, importUser, dataElementUid);
+          assertNull(changeLogs.get(0).change().dataValue().currentValue());
+          assertEquals("15", changeLogs.get(0).change().dataValue().previousValue());
```

Comment:
I might not fully get this case yet 😅 
The changelog 1 is for the initial create right?
The changelog 2 is for a delete and shows that no matter how often you delete a deleted value it will only lead to one change, right?